### PR TITLE
Reconcile extended resource capacity after kubelet restart.

### LINKED ISSUE #64632: From 1.11, an extended resource reported by a device plugin can be left on a node after node upgrade even though its device plugin never registers back
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>

/kind bug
> /kind feature

**What happened**:

In 1.11, https://github.com/kubernetes/kubernetes/pull/61877 removed externalID from Node.Spec which used to be used by kubelet to detect node recreation in certain environments during node upgrade. This changed node upgrade behavior in those environments because in pkg/kubelet/kubelet_node_status.go tryRegisterWithAPIServer(node *v1.Node), without externalID checking, we now consider the recreated node after upgrade as a previously registered node and thus will get old node state from API server instead of regenerating them based on node local state.

For extended resource reported by a device plugin, this causes problem because after node upgrade/recreation, the device plugin DaemonSet pod needs to be restarted by Kubelet, finishes certain setup, and then report to kubelet that its resource is available at the node. However, the node status capacity and allocatable from the API server still contains the old state before the node upgrade. After kubelet syncs up with the API server, the node status field gets overwritten and previously reported extended resources will appear on the node status capacity/allocatable even though the node is not ready for pods to consume such resources. As the result, Kubelet may start a pod requesting such resource without proper container runtime setup it needs to learn from device plugin.

Note to cope with the externalID removal change, cluster/gce/upgrade.sh has been updated (https://github.com/kubernetes/kubernetes/issues/63506) to explicitly uncordon the node after checking it is restarted and becomes ready. This however doesn't help device plugin case because node readiness doesn't depend on individual extended resource readiness.

This issue is to explore how we can fix this problem on head and 1.11.

In particular, I wonder whether we may switch to the model that we consider Kubelet as the only source to update extended resource capacity/allocatable in node status. I.e., don't support manually updating node status capacity/allocatable fields as documented in https://kubernetes.io/docs/tasks/administer-cluster/extended-resource-node/. This would allow us to re-generate node status capacity/allocatable every time in node status update, which I think is much simpler and more robust. As I heard, some folks have been using this mechanism to do simple resource exporting/accounting through a central controller. I wonder whether people are open to switch to the device plugin model in those cases. Even though those extended resources may not require special container setup, it seems a more secure model to have Kubelet totally own its node status.

FYI, here is the related OSS issue: https://github.com/kubernetes/kubernetes/issues/50473 where we first introduced extended resource concept.

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:
Running the gpu upgrade test from https://github.com/kubernetes/kubernetes/pull/63631 with --upgrade-target set to any 1.11 version would hit the issue.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by jiayingz**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes/kubernetes/issues/64632

**Special notes for your reviewer**:

**Release note**:

```release-note
Kubelet will set extended resource capacity to zero after it restarts. If the extended resource is exported by a device plugin, its capacity will change to a valid value after the device plugin re-connects with the Kubelet. If the extended resource is exported by an external component through direct node status capacity patching, the component should repatch the field after kubelet becomes ready again. During the time gap, pods previously assigned with such resources may fail kubelet admission but their controller should create new pods in response to such failures.
```

### CODE REVIEW

**Review by ConnorDoyle — COMMENTED** [review]
This seems to be a safe workaround to the problem described in #64632, especially with additional context provided by @vishh [here](https://github.com/kubernetes/kubernetes/issues/64632#issuecomment-394832599).

Is this true? With this patch, pods that consume extended resources can  no longer survive a Kubelet restart because they will fail admission. If so, could we add that to the release note?

File: pkg/kubelet/kubelet_node_status.go [hunk 1]
```diff
@@ -137,6 +138,18 @@ func (kl *Kubelet) tryRegisterWithAPIServer(node *v1.Node) bool {
 	return true
 }
 
+// Zeros out extended resource capacity during reconciliation.
+func (kl *Kubelet) reconcileExtendedResource(initialNode, node *v1.Node) bool {
+	requiresUpdate := false
+	for k := range node.Status.Capacity {
+		if v1helper.IsExtendedResourceName(k) {
+			node.Status.Capacity[k] = *resource.NewQuantity(int64(0), resource.DecimalSI)
```
[the remark this answers is not in the document]
Comment (jiayingz) [reply pkg/kubelet/kubelet_node_status.go]: Done.

**Comment by jiayingz** [comment]
@ConnorDoyle Thanks a lot for the comment! I updated the release note to clarify the pod failure behavior associated with the change. PTAL.

File: pkg/kubelet/kubelet_node_status_test.go [hunk 2]
```diff
@@ -1526,6 +1526,80 @@ func TestUpdateDefaultLabels(t *testing.T) {
...
+	cases := []struct {
+		name         string
+		existingNode *v1.Node
+		expectedNode *v1.Node
+		needsUpdate  bool
+	}{
+		{
+			name: "no update needed without extended resource",
+			existingNode: &v1.Node{
+				Status: v1.NodeStatus{
```
Comment (RenaudWasTaken) [inline pkg/kubelet/kubelet_node_status_test.go]: Should this test check Allocatable?

Comment (jiayingz) [reply pkg/kubelet/kubelet_node_status_test.go]: Ah my bad. I did extend the test to check allocatable but forgot to push the change.

Comment (RenaudWasTaken) [reply pkg/kubelet/kubelet_node_status_test.go]: Ah you just updated it!

**Comment by vishh** [comment]
@jiayingz what will be the behavior after this PR for trivial kubelet restarts? Will the kubelet update capacity/allocatable even though it has a valid checkpoint?

**Comment by jiayingz** [comment]
For device plugin resource, even with a valid checkpoint, we already sets the resource capacity/allocatable to zero since https://github.com/kubernetes/kubernetes/pull/60856 to make sure no new pods will get assigned to the node till the device plugin re-connects. Existing pods already assigned with the resource can continue though with the valid checkpoint in place. That is the logic covered from device manager Allocate().

**Pull request merged**